zio/jsonio/writer.go

```diff
@@ -1,26 +1,250 @@
 package jsonio
 
 import (
+	"bufio"
+	"bytes"
+	"encoding/hex"
 	"encoding/json"
+	"fmt"
 	"io"
+	"time"
 
 	"github.com/brimdata/zed"
+	"github.com/brimdata/zed/pkg/nano"
+	"github.com/brimdata/zed/pkg/terminal/color"
+	"github.com/brimdata/zed/zcode"
+	"github.com/brimdata/zed/zson"
+)
+
+var (
+	boolColor   = []byte("\x1b[1m")
+	fieldColor  = []byte("\x1b[34;1m")
+	nullColor   = []byte("\x1b[2m")
+	numberColor = []byte("\x1b[36m")
+	puncColor   = []byte{} // no color
+	stringColor = []byte("\x1b[32m")
 )
 
 type Writer struct {
 	io.Closer
-	encoder *json.Encoder
+	writer *bufio.Writer
+	tab    int
+
+	// Use json.Encoder to get marshal primitive Values. Have to use
+	// json.Encoder instead of json.Marshal because this is the only way to get
+	// pkg json to turn off HTML escaping.
+	primEnc *json.Encoder
+	primBuf bytes.Buffer
 }
 
-func NewWriter(wc io.WriteCloser) *Writer {
-	e := json.NewEncoder(wc)
-	e.SetEscapeHTML(false)
-	return &Writer{
-		Closer:  wc,
-		encoder: e,
+type WriterOpts struct {
+	Pretty int
+}
+
+func NewWriter(writer io.WriteCloser, opts WriterOpts) *Writer {
+	w := &Writer{
+		Closer: writer,
+		writer: bufio.NewWriter(writer),
+		tab:    opts.Pretty,
 	}
+	w.primEnc = json.NewEncoder(&w.primBuf)
+	w.primEnc.SetEscapeHTML(false)
+	return w
 }
 
 func (w *Writer) Write(val zed.Value) error {
-	return w.encoder.Encode(marshalAny(val.Type(), val.Bytes()))
+	w.marshal(0, val)
+	w.writer.WriteByte('\n')
+	return w.writer.Flush()
+}
+
+func (w *Writer) marshal(tab int, val zed.Value) {
```

Comment:
Nit: This no longer feels like the right name so I'd do s/marshal/write/g in this file.